### PR TITLE
Use CakeBuild for building osu!

### PR DESCRIPTION
**Pull request by miterosan**
Use cakebuild for building, testing and analyzing the osu! source code.

This allows running the code checks locally.

~Please note that the code file sanity tool and NVika do not have bindings for cakebuild yet.~

### CODE REVIEW

File: build.cake [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+#tool "nuget:?package=NVika.MSBuild"
+var NVikaToolPath = GetFiles("./tools/NVika.MSBuild.*/tools/NVika.exe").First();
+var CodeFileSanityToolPath = DownloadFile("https://github.com/peppy/CodeFileSanity/releases/download/v0.2.5/CodeFileSanity.exe");
+
+///////////////////////////////////////////////////////////////////////////////
+// ARGUMENTS
+///////////////////////////////////////////////////////////////////////////////
+
+var target = Argument("target", "Build");
+var framework = Argument("framework", "net471");
```
Comment (VPeruS) [inline build.cake]: Desktop has netcoreapp2.1 as target framework

Comment (miterosan) [reply build.cake]: This is true, but it was always build with net471 in the appveyor.xml. @peppy can we switch over to netcoreapp2.1 or do we stay on net471 for now?

Comment (VPeruS) [reply build.cake]: appveyor use value from csproj -> netcoreapp2.1

Comment (miterosan) [reply build.cake]: nevermind then

Comment (peppy) [reply build.cake]: We made the complete switch yesterday via https://github.com/ppy/osu/pull/3061

**Comment by swoolcock** [comment]
Will this work for Xamarin?

**Comment by peppy** [comment]
It's just a script engine; it can be made to work with anything.

**Comment by peppy** [comment]
Can you update this to consume CFS via nuget? See https://github.com/ppy/osu/pull/3137

**Comment by miterosan** [comment]
.-.  I am against having the end user installing codefilesanity globally on his pc. The local build should not require random programs preinstalled. The point of switching to cakebuild was that the developers can test and validate the code by just running build.ps1/build.sh. Installing CodeFileSanity as a global tool would defeat that purpose. 

Dotnetcoretools are supposed to be added as a packagereference and then called over dotnet somedotnettool osu.desktop.csproj. You can not use a dotnettool without 

The way we can go about this is implementing the cake addin alias interface. This allows customizability of codefilesanity inside the cake build script.
https://codeopinion.com/how-to-create-cake-addin/

TLDR:
- dotnetcore Global tools require installing
- dotnetcore tools can only appliend on single csproj files that packagereference the tool.
- implementing the cakealias interface is clean and customizable.

^ I am currently on this.

**Comment by peppy** [comment]
Oh yeah, no shit. I am 100% with you on that one. I even explained to smoogi the same thing over a good 30 minutes. The global install is fine in the appveyor config because it's a once-off run, but doing that locally would be blasphemy.

**Comment by miterosan** [comment]
This build needs to be repeated when https://github.com/peppy/CodeFileSanity/pull/5 is merged and deployed.

File: build.cake [hunk 2]
```diff
@@ -38,17 +38,14 @@ Task("Test")
     });
 });
 
+// windows only because both inspectcore and nvika depend on net45
 Task("InspectCode")
+.WithCriteria(IsRunningOnWindows())
 .IsDependentOn("Compile")
 .Does(() => {
     var nVikaToolPath = GetFiles("./tools/NVika.MSBuild.*/tools/NVika.exe").First();
     var nugetToolPath = GetFiles("./tools/NuGet.CommandLine.*/tools/NuGet.exe").First();
```
Comment (peppy) [inline build.cake]: `dotnet build` automatically does a nuget restore. this is unnecessary.

Comment (miterosan) [reply build.cake]: The tests are not restored at that point of time (this runs before the tests). Windows nuget makes no sense here, yes.

**Comment by peppy** [comment]
@miterosan what's the status on this one?

**Comment by miterosan** [comment]
@peppy Waiting that you merge/review peppy/CodeFileSanity#5. This and the codeFileSanity are ready on my side.

**Comment by peppy** [comment]
oh, sorry. thought i already got that one in...

**Comment by miterosan** [comment]
Bout this issue. This should be ready. Please note that you need to create a release of the codefilesanity one for this to make sense.

**Comment by peppy** [comment]
Should already be published: https://www.nuget.org/packages/CodeFileSanity/0.0.21

**Comment by miterosan** [comment]
I pinned the tool and addin versions down to allow repeatable builds.
The is PR is mergeable from my side when the CI goes green.

**Comment by miterosan** [comment]
for reference: why is the build green if there are failing tests?

**Comment by peppy** [comment]
The exit code is probably 0?

**Comment by miterosan** [comment]
So... this should be ready for review again.

**Comment by miterosan** [comment]
Is ready for review.

I increased the used cakebuild version to the latest.
Tested the buildscript on both windows 1809 and on lubuntu (Ubuntu 18.04.1 LTS x84-64, linux powershell version v6.1.0-preview.2, dotnet version 2.1.302)

**Comment by jai-x** [comment]
I would suggest a squash and rebase into one commit with a properly written commit message.

**Comment by miterosan** [comment]
can do when approval is given

**Comment by smoogipoo** [comment]
On macOS: `Error: Failed to install tool 'NVika.MSBuild'.`

**Comment by peppy** [comment]
The readme should probably be updated with build/test instructions using this method.

**Comment by peppy** [comment]
Is there a reason for only providing a powershell bootstrap for this? May be useful to also have a bash script.

**Comment by peppy** [comment]
```PS C:\Users\Dean\Projects\osu> .\build.ps1
.\build.ps1 : File C:\Users\Dean\Projects\osu\build.ps1 cannot be loaded because running scripts is disabled on this
system. For more information, see about_Execution_Policies at https:/go.microsoft.com/fwlink/?LinkID=135170.
At line:1 char:1
+ .\build.ps1
+ ~~~~~~~~~~~
    + CategoryInfo          : SecurityError: (:) [], PSSecurityException
    + FullyQualifiedErrorId : UnauthorizedAccess```

is this normal for a vanilla windows install?

**Comment by goatdavid** [comment]
yeah, iirc you need to run powershell as admin and use Set-ExecutionPolicy
RemoteSigned

2018年10月29日(月) 午前10:37 Dean Herbert <notifications@github.com>:

**Comment by peppy** [comment]
Works as expected on mac and windows. Let's give this a try.

**Review by peppy — APPROVED** [review]
